### PR TITLE
Integrated activation behaviour into event dispatch algorithm

### LINKED ISSUE #22783: Integrate activation behaviour into event dispatch algorithm
`synthetic_click_activation` implements an old algorithm for activating elements via synthetic mouse click events. We should rewrite it to implement [this synthetic mouse event algorithm](https://html.spec.whatwg.org/multipage/#fire-a-synthetic-mouse-event) and integrate the activation behaviour into the event dispatch algorithm [per the spec](https://dom.spec.whatwg.org/#concept-event-dispatch). This will address the comment in `Document::dispatch_key_event` about synthetic click activiation being unspecified, since it has been specified appropriately in the past 4 years.

Code: `components/script/dom/document.rs`, `components/script/dom/eventtarget.rs`, `components/script/activation.rs`

### PR DESCRIPTION
**Pull request by aditj**
Trying to update the synthetic mouse click algorithm according to the [spec](https://html.spec.whatwg.org/multipage/#fire-a-synthetic-mouse-event)

---

- [ ] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #22783  (GitHub issue number if applicable)

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22981)

### CODE REVIEW

**Comment by aditj** [comment]
Help Please! How can I remove the previously approved commit from this pull? :D

**Comment by jdm** [comment]
I recommend doing this:
```
$ git reset --hard de09c8fb6ef87dec5932d5fab4adcb421d291a54
$ git cherry-pick 009db7396362dc4febba4f5bb43d08cd0e6feb9e
$ git push -f origin aditj-patch-6
```

File: components/script/dom/activation.rs [hunk 1]
```diff
@@ -99,13 +88,14 @@ pub fn synthetic_click_activation(
...
     let event = mouse.upcast::<Event>();
-    if source == ActivationSource::FromClick {
+    if not_trusted {
         event.set_trusted(false);
     }
+    //Step 9
     target.dispatch_event(event);
-
-    // Step 5
+    event.dispatching();
```
Comment (Manishearth) [inline components/script/dom/activation.rs]: this does nothing, please remove this line

the spec asks us to return the result of dispatching, that would mean returning the `EventStatus` returned  by `dispatch_event` (however I don't see anything in the spec that uses this right now)

File: components/script/dom/event.rs [hunk 2]
```diff
@@ -169,6 +170,9 @@ impl Event {
         // Step 3-4.
         let path = self.construct_event_path(&target);
         rooted_vec!(let event_path <- path.into_iter());
+        if self.type_() == atom!("click") {
+            is_activation_event = true;
```
Comment (Manishearth) [inline components/script/dom/event.rs]: This variable isn't used, also you can fire manual click events that aren't activation events. Please remove the is_activation_event stuff.

File: components/script/dom/activation.rs [hunk 3]
```diff
@@ -48,43 +48,32 @@ pub trait Activatable {
...
-    if element.click_in_progress() {
-        return;
-    }
-    // Step 2
-    element.set_click_in_progress(true);
-    // Step 3
+    // Step 12 of https://dom.spec.whatwg.org/#concept-event-dispatch
     let activatable = element.as_maybe_activatable();
     if let Some(a) = activatable {
         a.pre_click_activation();
```
Comment (Manishearth) [inline components/script/dom/activation.rs]: As mentioned in [@jdm's comment](https://github.com/servo/servo/issues/22783#issuecomment-469793779), this stuff needs to be moved into the event dispatch code

https://dom.spec.whatwg.org/#concept-event-dispatch

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: components/script/dom/event.rs (hunk 2 above)
Comment (aditj) [reply components/script/dom/event.rs]: I think is_activition_event will be used in Steps 9.6.1 and Steps 9.8.1 of the [spec](https://dom.spec.whatwg.org/#concept-event-dispatch)

Comment (Manishearth) [reply components/script/dom/event.rs]: Ah, I see. Can you implement that part?

File: components/script/dom/activation.rs (hunk 3 above)
Comment (aditj) [reply components/script/dom/activation.rs]: I'll remove this from here.
Coming to implementation in the event dispatch code there's a [comment](https://github.com/servo/servo/blob/d0ac1742bbe9425e8418c0884d461fd786aba35c/components/script/dom/event.rs#L187) mentioning step 12 of the [spec](https://dom.spec.whatwg.org/#concept-event-dispatch) . Although I am unsure of the correctness of this [implementation](https://github.com/servo/servo/blob/d0ac1742bbe9425e8418c0884d461fd786aba35c/components/script/dom/event.rs#L208).

File: components/script/dom/event.rs (hunk 2 above)
Comment (aditj) [reply components/script/dom/event.rs]: I can certainly try!

**Comment by jdm** [comment]
@Manishearth Re-review ping.

**Comment by Manishearth** [comment]
@jdm The requested changes haven't yet been made

File: components/script/dom/event.rs (hunk 2 above)
Comment (Manishearth) [reply components/script/dom/event.rs]: Downcast it to an element, and if that succeeds, call `.as_maybe_activatable()` on it.

**Comment by jdm** [comment]
Closing due to inactivity.

**Pull request closed without merging**